pkg/ruby/code_search.go

```diff
@@ -0,0 +1,179 @@
+// is used to calculate the corresponding cache file.
+func (o *Options) defaultSHA(pkg *Package) (string, error) {
+	ctx := context.Background()
+	client := github.NewClient(o.Client.Client)
+
+	// Get the repository
+	repository, _, err := client.Repositories.Get(ctx, pkg.Repo.Organisation, pkg.Repo.Name)
+	if err != nil {
+		return "", err
+	}
+
+	// Get the main branch reference
+	reference, _, err := client.Git.GetRef(ctx, pkg.Repo.Organisation, pkg.Repo.Name, "heads/"+*repository.DefaultBranch)
+	if err != nil {
+		return "", err
+	}
+
+	// Get the commit SHA of the main branch
+	return *reference.Object.SHA, nil
+}
+
+// cachedSearchResult returns the path for a cache file
+func (o *Options) cachedSearchResult(pkg *Package, sha, query string) string {
+	return path.Join(rubyCacheDirectory, pkg.Name, fmt.Sprintf("%s-%s.json", sha, url.QueryEscape(query)))
+}
+
+// runQuery performs a code search using the Github API given a particular
+// query string. It will append a repo:xxx/yyy to only search one repo at a
+// time for a particular query string.
+//
+// A note on caching:
+//
+// Github limits searches even further than normal API calls. To get around
+// this (mainly to make dev easier) the raw json result is cached to make it
+// easy to probe further without hitting the Github API every single time.
+// Because Github Code Search limits searching to the default branch only, the
+// cached file is named with the latest commit sha on the default branch. That
+// way the cache does not get stale if new changes are made to the repository.
+func (o *Options) runQuery(pkg *Package, query string) ([]SearchResult, error) {
+	logger := log.New(log.Writer(), "wolfictl ruby code-search: ", log.LstdFlags|log.Lmsgprefix)
+	sha, err := o.defaultSHA(pkg)
+	if err != nil {
+		return nil, fmt.Errorf("getting default branch sha: %w", err)
+	}
+
+	cachedPath := o.cachedSearchResult(pkg, sha, query)
+	cached, err := os.Open(cachedPath)
+	if err != nil || o.NoCache {
+		ctx := context.Background()
+
+		client := github.NewClient(o.Client.Client)
+		query = fmt.Sprintf("%s repo:%s/%s", query, pkg.Repo.Organisation, pkg.Repo.Name)
+		gitOpts := gh.GitOptions{
+			GithubClient: client,
+			Logger:       logger,
+		}
+
+		result, err := gitOpts.SearchCode(ctx, query)
+		if err != nil {
+			fmt.Printf("Error: %+v\n", err)
+			return nil, err
+		}
+
+		err = os.MkdirAll(path.Dir(cachedPath), 0o755)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create cache directory: %w", err)
+		}
+
+		cached, err = os.Create(cachedPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create cache file: %w", err)
+		}
+
+		// Convert the struct to JSON
+		jsonData, err := json.Marshal(result)
+		if err != nil {
+			return nil, fmt.Errorf("marshaling json: %w", err)
+		}
+
+		_, err = cached.Write(jsonData)
+		if err != nil {
+			return nil, fmt.Errorf("failed to write cache file: %w", err)
+		}
+
+		_, err = cached.Seek(0, io.SeekStart)
```

Comment:
might be worth hoisting this leg out into a function so you can defer the closing without needing to rewind

Reply:
Good note, did the hoisting. Happy to have any feedback :)